pingcap/tidb-operator
# Support IAM role for backup CRD

### What problem does this PR solve? 
Support using backup CRD in kube2iam environment
### What is changed and how does it work?
If the backup CRD has an iam annotation, ignore check the accesskey and secretkey and annotate the iam annotation to the backup job
### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

use the kube2iam env
create a cluster, annotation tikv pod with annotation:
```
iam.amazonaws.com/role: "{IAMROLE}"
```
PS: "{IAMROLE}" is the used iam

create backup with backup yaml 
```
apiVersion: pingcap.com/v1alpha1
kind: Backup
metadata:
  name: backuptest1
  namespace: tidb1
  annotations: 
    iam.amazonaws.com/role: "{IAMROLE}"
spec:
  br:
    enableTLSClient: true
    cluster: db
    clusterNamespce: tidb1
    enableTLSClient:  true
    sendCredToTikv:  false
  from:
    host:         172.30.6.56
    port:         4000
    secretName:   mysql-pwd
    user:         root
  s3:
    provider: aws
    region: us-west-2
    bucket: my-bucket
    prefix: backup-test
  storageClassName: local-storage
  storageSize: 10Gi
```
check the aws backup storage with the backup file

In the same env
create restore with restore.yaml
```
apiVersion: pingcap.com/v1alpha1
kind: Restore
metadata:
  name: restoretest1
  namespace: tidb1
  annotations: 
    iam.amazonaws.com/role: "{IAMROLE}"
spec:
  br:
    enableTLSClient: true
    cluster: db
    clusterNamespce: tidb1
    enableTLSClient:  true
    sendCredToTikv:  false
  to:
    host:         172.30.6.56
    port:         4000
    secretName:   mysql-pwd
    user:         root
  s3:
    provider: aws
    region: us-west-2
    bucket: my-bucket
    prefix: backup-test
  storageClassName: local-storage
  storageSize: 10Gi
```
check the sql of the number of data

Code changes

 - Has Go code change

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
The `spec.s3.secretName` field is not required any more in `Backup`, `Restore` and `BackupSchedule` custome resources.
Users can assign an IAM role to the backup or restore Pod by setting `metadata.annotations` or `spec.serviceAccount` in the custom resources.
```
